h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TEBroadcastJoin.java

```diff
@@ -0,0 +1,218 @@
+
+class TEBroadcastJoin {
+
+  /**
+   * @param leftFrame frame for which we want to keep the rows order.
+   * @param leftCatColumnsIdxs indices of the categorical columns from `leftFrame` for which we want to calculate encodings.
+   *        Only one categorical column is currently supported.
+   * @param leftFoldColumnIdx index of the fold column from `leftFrame` or `-1` if we don't use folds.
+   * @param rightFrame supposedly small frame that we will broadcast to all nodes and use as a lookup table for joining.
+   * @param rightCatColumnsIdxs indices of the categorical columns from `rightFrame`.
+   *        Only one categorical column is currently supported.
+   * @param rightFoldColumnIdx index of the fold column from `rightFrame` or `-1` if we don't use folds.
+   * @param maxFoldValue the highest fold value (e.g. 4 if 5 folds).
+   * @return the `leftFrame` with joined numerators and denominators.
+   */
+  static Frame join(Frame leftFrame, int[] leftCatColumnsIdxs, int leftFoldColumnIdx,
+                    Frame rightFrame, int[] rightCatColumnsIdxs, int rightFoldColumnIdx,
+                    int maxFoldValue) {
+    int rightNumeratorIdx = rightFrame.find(TargetEncoderHelper.NUMERATOR_COL);
+    int rightDenominatorIdx = rightNumeratorIdx + 1; // enforced by 
+    
+    // currently supporting only one categorical column
+    assert leftCatColumnsIdxs.length == 1;
+    assert rightCatColumnsIdxs.length == 1;
+    int leftCatColumnIdx = leftCatColumnsIdxs[0];
+    int rightCatColumnIdx = rightCatColumnsIdxs[0];
+    int rightCatCardinality = rightFrame.vec(rightCatColumnIdx).cardinality();
+
+    if (rightFoldColumnIdx != -1 && rightFrame.vec(rightFoldColumnIdx).max() > Integer.MAX_VALUE)
+      throw new IllegalArgumentException("Fold value should be a non-negative integer (i.e. should belong to [0, Integer.MAX_VALUE] range)");
+      
+    int[][] levelMappings = {
+            CategoricalWrappedVec.computeMap(
+                    leftFrame.vec(leftCatColumnIdx).domain(), 
+                    rightFrame.vec(rightCatColumnIdx).domain()
+            )
+    };
+
+    double[][] encodingData = encodingsToArray(
+            rightFrame, rightCatColumnIdx, rightFoldColumnIdx, 
+            rightNumeratorIdx, rightDenominatorIdx, 
+            rightCatCardinality, maxFoldValue
+    );
+
+    // BroadcastJoiner is currently modifying the frame in-place, so we need to provide the numerator and denominator columns.
+    Frame resultFrame = new Frame(leftFrame);
+    resultFrame.add(TargetEncoderHelper.NUMERATOR_COL, resultFrame.anyVec().makeCon(0));
+    resultFrame.add(TargetEncoderHelper.DENOMINATOR_COL, resultFrame.anyVec().makeCon(0));
+    new BroadcastJoiner(leftCatColumnsIdxs, leftFoldColumnIdx, encodingData, levelMappings, rightCatCardinality-1)
+            .doAll(resultFrame);
+    return resultFrame;
+  }
+  
+  static double[][] encodingsToArray(Frame encodingsFrame, int categoricalColIdx, int foldColIdx, 
```

Comment:
TBH I don't understand the point of the original design (I fully realize this PR can't resolve evertything) moving everything back and forth between frame representation and arrays. I'd personally try to avoid these steps. Just a personal note.

Reply:
It confused me as well, don't want to change this here though: it works. I resisted in not changing a few things to avoid getting sunk into this :)